Merge all features related to C# 7 ref features (#795)
dotnet/csharpstandard: standard/variables.md

### 9.2.1 General

> C# defines eight categories of variables: static variables, instance variables, array elements, value parameters, input parameters, reference parameters, output parameters, and local variables. The subclauses that follow describe each of these categories.

Reviewer: ```suggestion
C# defines nine categories of variables: static variables, instance variables, array elements, value parameters, input parameters, reference parameters, output parameters, local variables, [and an almost fanatical devotion to the Pope](https://youtu.be/Cj8n4MfhjUc?t=47). The subclauses that follow describe each of these categories.
```

Reviewer: (I'm trusting @BillWagner not to actually commit this...)

Author: I am so tempted....

A parameter declared with a `ref` modifier is a ***reference parameter***.

> A reference parameter refers to the variable given as the argument in the function member, anonymous function, or local function invocation. The variable given as the argument is the ***referent*** of the reference parameter. Thus, the value of a reference parameter is always the same as the underlying variable.

Reviewer 2: We have duplication between the clauses here and those on parameters in in classes.md (§15.6.2.3 and friends), I will not attempt to address that. Instead I'll just suggest corresponding changes to those made in classes.md, except I'm not adding to the duplication by adding notes that the referent can be changed.

```suggestion
A reference parameter is a reference variable (§ref-span-safety) which comes into existence upon invocation of the function member, delegate, anonymous function, or local function and its referent is initialized to the variable given as the argument in that invocation. A reference parameter ceases to exist when execution of the function body completes. Unlike value parameters a reference parameter may not be captured (§ref-span-safety-limitations).
```

A parameter declared with an `out` modifier is an ***output parameter***.

> An output parameter refers to the variable given as the argument in the function member or delegate invocation. The variable given as the argument is the referent of the output parameter. Thus, the value of an output parameter is always the same as the underlying variable.

Reviewer 2: Following reference parameter above:

```suggestion
An output parameter is a reference variable (§ref-span-safety) which comes into existence upon invocation of the function member, delegate, anonymous function, or local function and its referent is initialized to the variable given as the argument in that invocation. An output parameter ceases to exist when execution of the function body completes. Unlike value parameters an output parameter may not be captured (§ref-span-safety-limitations).
```

- Within a function member or anonymous function, an output parameter is considered initially unassigned.
- Every output parameter of a function member, anonymous function, or local function shall be definitely assigned ([§9.4](variables.md#94-definite-assignment)) before the function member, anonymous function, or local function returns normally.

### §input-parameters-new-clause Input parameters

A parameter declared with an `in` modifier is an ***input parameter***.

> An input parameter refers to the variable given as the argument in the function member or delegate invocation. The variable given as the argument is the referent of the input parameter. Thus, the value of an input parameter is always the same as the underlying variable.

Reviewer 2: Following reference & output parameters above:

```suggestion
An input parameter is a reference variable (§ref-span-safety) which comes into existence upon invocation of the function member, delegate, anonymous function, or local function and its referent is initialized to the variable given as the argument in that invocation. An input parameter ceases to exist when execution of the function body completes. Unlike value parameters an input parameter may not be captured (§ref-span-safety-limitations).
```

A parameter declared with an `in` modifier is an ***input parameter***.

> An input parameter is a reference variable (§ref-span-safety) which comes into existence upon invocation of the function member, delegate, anonymous function, or local function and its referent is initialized to the variable given as the argument in that invocation. An input parameter ceases to exist when execution of the function body completes. Unlike value parameters an input parameter may not be captured (§ref-span-safety-limitations).

Reviewer 2: > its referent is initialized to the variable given as the argument in that invocation

Technically correct as a variable reference is always the argument, but in some cases it is a reference to a compiler introduced temporary – see invocation details.

I wouldn't try to incorporate that directly here, but those who like *Note*s might like to consider whether it's worth adding one.

I will however suggest we change the use of *variable*:

```suggestion
An input parameter is a reference variable (§ref-span-safety) which comes into existence upon invocation of the function member, delegate, anonymous function, or local function and its referent is initialized to the *variable_reference* given as the argument in that invocation. An input parameter ceases to exist when execution of the function body completes. Unlike value parameters an input parameter may not be captured (§ref-span-safety-limitations).
```

Reviewer 2: @BillWagner – I might have confused here by making a remark about a Note _and_ making a suggested fix to the text?

Author: Unresolving to discuss if we want to add a note.

Reviewer 2: The 👍🏼 is to let the group discuss – one person's valued note is another's unnecessary addition ;-) In this case I wouldn't myself add one, but if someone wishes to propose one go for it!

## §ref-span-safety Reference variables and returns

### §ref-span-safety-general General

A ***reference variable*** is a variable that refers to another variable, called the referent (§9.2.6). A reference variable does not store the value of its referent. When a reference variable is used where a value is required its referent's value is returned; similarly when a reference variable is the target of an assignment it is the referent which is assigned to. The variable to which a reference variable refers, i.e. its referent, can be changed using a ref assignment (`= ref`). A reference variable is a local variable declared with the `ref` modifier.

> *Example:* The following example demonstrates a local reference variable whose referent is an element of an array:
>
> ```csharp
> public class C
> {
>
>     public void M()
>     {
>         int[] arr = new int[10];
>         // element is a reference variable that refers to arr[5]
>         ref int element = ref arr[5];
>         element += 5; // arr[5] has been incremented by 5
>     }     
> }
> ```
>
> *end example*

> A ***reference return*** is the expression returned by reference from a method whose return type includes the `ref` or `ref readonly` modifiers (§15.6.1). The variable expression of the reference return is the referent of the reference return.

Reviewer 2: Aligning terminology between here and §15.6.1 & §9.6. There already exists a term for what is stored in a reference variable, *variable_reference* (§9.6). However the language used for `ref`, `out` and now `in` parameters has studiously hidden this; which is not an invalid approach *per se* and some languages do this; but with the introduction of reference variables and returns in C# they become visible.

Here I'll unify the description of reference variables here and that of reference returns in §15.6.1. As the latter already refers to *variable_reference* it is being used as the base for updating here.

```suggestion
A ***reference variable*** is a variable that refers to another variable, called the referent (§9.2.6).  A reference variable is a local variable declared with the `ref` modifier.

A reference variable stores a *variable_reference* (§9.6) to its referent and not the value of its referent. When a reference variable is used where a value is required its referent's value is returned; similarly when a reference variable is the target of an assignment it is the referent which is assigned to. The variable to which a reference variable refers, i.e. the stored *variable_reference* for its referent, can be changed using a ref assignment (`= ref`).

> *Example:* The following example demonstrates a local reference variable whose referent is an element of an array:
>
> ```csharp
> public class C
> {
>
>     public void M()
>     {
>         int[] arr = new int[10];
>         // element is a reference variable that refers to arr[5]
>         ref int element = ref arr[5];
>         element += 5; // arr[5] has been incremented by 5
>     }     
> }
> ```
>
> *end example*

A ***reference return*** is the *variable_reference* returned from a returns-by-ref method (§15.6.1). This *variable_reference* is the referent of the reference return.
```

> *Example:* The following example demonstrates a reference return whose referent is an element of an array field:
>
> ```csharp
> public class C
> {
>     private int[] arr = new int[10];
>
>     public readonly ref int M()
>     {
>         // element is a reference variable that refers to arr[5]
>         ref int element = ref arr[5];
>         return ref element; // return reference to arr[5];
>     }     
> }
> ```
>
> *end example*

### §ref-safe-contexts Ref safe contexts

All reference variables obey safety rules that ensure the context of the reference variable is not greater than the ref-safe-context of its referent.

> For any variable, the ***ref-safe-context*** of that variable is the context where a *variable_reference* (§9.5) to that variable is valid. The referent of a reference variable must have a ref-safe-context that is at least as wide as the context of the reference variable.

Reviewer 2: Reading this doesn't flow well to me, maybe because an unadorned “context” has other (English & computer) meanings. Maybe:

```suggestion
All reference variables obey safety rules that ensure the ref-safe-context of the reference variable is not greater than the ref-safe-context of its referent.

For any variable, the ***ref-safe-context*** of that variable is the context where a *variable_reference* (§9.5) to that variable is valid. The referent of a reference variable must have a ref-safe-context that is at least as wide as the ref-safe-context of the reference variable itself.
```

> *Note*: The compiler determines the safe context through a static analysis of the program text. The safe context reflects the lifetime of a variable at runtime. *end note*

> There are three valid ref-safe-contexts:

Reviewer 2: I don't think we have *invalid* ref-safe-contexts *per se* do we? Defined maybe? Or just drop the word:

```suggestion
There are three ref-safe-contexts:
```

> - ***declaration-block***: A *variable_reference* to a local variable declared in a block is valid from its declaration to the end of the block in which it is declared. Each nested block represents a different context.  The ref-safe-context of a local variable is the declaration-block in which it is declared. A local variable declared in a method has a ref-safe-context of the block that defines the method. A variable declared in a block is a valid referent only if the reference variable is declared in the same block after the referent, or a nested block.

Reviewer 2: Not a fully formed suggestion, may contain E&O:

> ***declaration-block***: The ref-safe-scope of *variable_reference* to a local variable ([§9.2.8](variables.md#928-local-variables)) extends from its declaration to the end of the scope in which it is declared, including into any nested *embedded-statement*s in this range.
>
> A *variable_reference* to a local variable is a valid referent for a reference variable only if the reference variable is declared within the ref-safe-scope of that variable.

Open issues:
- Relies on using *scope*, as used in §9.2.8, but need to check how well-defined that is.
- Effectively equates *scope* and *ref-safe-scope* for local variables, is this correct?

Author: I did check the text of §9.2.8, and that is well-defined enough for this purpose. Yes, equating the *scope* of a local variable with its *ref-safe-context* is correct.

> - ***function-member***: A *variable_reference* to a value parameter on a function member declaration, including the implicit `this` parameter, is valid in the entire function member. The ref-safe-context of the fields of a `struct` type is function-member. A variable with ref-safe-context of function-member is a valid referent only if the reference variable is declared in the same function member.

Reviewer 2: What does “The ref-safe-context of the fields of a `struct` type is function-member” mean? Without this sentence this paragraph states that parameters can only be referents for reference variables declared within the function member – all good.

Is not the ref-safe-context of a field the the same as the ref-safe-context of whole struct, and that struct value must be declared somewhere and its ref-safe-scope is therefore declaration-block? I.e. is the ref-safe-context of a part ever different than the whole?

Author: The reason is because the `this` parameter for a `struct` method is passed-by-value. So, `this` is *a copy* of the `struct` passed as an argument.

Unless what you mean is to remove "entire" from the previous sentence?

Reviewer 3: > The reason is because the `this` parameter for a `struct` method is passed-by-value.

The `this` parameter in a struct instance method is a `ref`. Did I misunderstand what this sentence was referring to?

```cs
var s = new S { F = 0 };
s.M();
Console.WriteLine(s.F); // 1

struct S
{
    public int F;
    public void M() { F = 1; }
}
```

Author: Hmm I may be confused.  Why does this variation of `S` not compile? 

```
struct S
{
    public int F;
    public void M() { F = 1; }
    public ref int GetF() => ref F; // Error: Struct members cannot return 'this' or other instance members by reference
}
```

If `this` is a ref parameter, shouldn't the fields of `this` be *safe-to-ref-return*?

Reviewer 3: It is a ref parameter, but a special one with special rules in this version. It is not returnable. In C# 11 the `this` parameter in struct methods is equivalent to a `scoped ref` parameter. (In other words, in C# 11 we added the ability to declare parameters which have equivalent ref safety behavior as the `this` parameter.)

Author: From the discussion, the spec language is correct. I added a clarifying note.

> - ***caller-context***: Member fields of a `class` type and `ref` parameters have ref-safe-context of caller-context. A variable with ref-safe-context of caller-context can be the referent of a reference return. A variable that can be the referent of a reference-return is ***safe-to-ref-return***.

Reviewer 2: I think this says that within a method a new object can be allocated, its fields will have caller-context, and so a reference to one of its accessible fields can be returned from the method, and that could be the only extant reference to the object (or part of it) after the return…

I doubt that is what is intended?

So consider the class `C` declared in the example below but with the accessibility of `arr` changed to `public`:

```csharp
public class C
 {
     // ref safe context of arr is "caller-context". 
     // ref safe context of arr[i] is "caller-context".
     public int[] arr = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 };
```

Now a method *not* a member of `C`:

```csharp
     public ref int NotInC()
     {
         C localObject = new C();
 
         // the ref-safe-context of localObject, and hence its members,
         // is from its declaration to the end of the method, yet the above
         // para seems to say its members are safe-to-return…
         return ref localObject.arr[v2]; // invalid surely (unless a reference to part keeps the whole alive)?
     }
```
*I won't review the rest of the changes stemming from #742 for now*

Author: That isn't what was meant. I think the change is as follows:

```suggestion
- ***caller-context***: Member fields of a `class` type parameter, including the receiver, and `ref` parameters have ref-safe-context of caller-context. A variable with ref-safe-context of caller-context can be the referent of a reference return. A variable that can be the referent of a reference-return is ***safe-to-ref-return***.
```

Reviewer 2: @BillWagner – Duh, I should have figured out that reading! Thanks, change looks good.

> - ***caller-context***: Member fields of a `class` type and `ref` parameters have ref-safe-context of caller-context. A variable with ref-safe-context of caller-context can be the referent of a reference return. A variable that can be the referent of a reference-return is ***safe-to-ref-return***.

Reviewer 2: placeholder…

> - ***function-member***: A *variable_reference* to a value parameter on a function member declaration, including the implicit `this` parameter, is valid in the entire function member. The ref-safe-context of the fields of a `struct` type is function-member. A variable with ref-safe-context of function-member is a valid referent only if the reference variable is declared in the same function member.
> - ***caller-context***: Member fields of a `class` type and `ref` parameters have ref-safe-context of caller-context. A variable with ref-safe-context of caller-context can be the referent of a reference return. A variable that can be the referent of a reference-return is ***safe-to-ref-return***.

Reviewer 2: A comment thread by @BillWagner & @RikkiGibson prompted some archeology in the [bowels of Roslyn](https://github.com/dotnet/roslyn/issues/5233) which uncovered the intended semantics for the context of the implicit `this` (a `ref` parameter) in a `struct` instance method: `this` (and its fields) have function-member context and are **not** safe-to-ref-return like other `ref` parameters.

The following suggestion incorporates these semantics into function-member and caller-context. Also added is text for arrays. The
suggestion includes @BillWagner’s [clarifications](https://github.com/dotnet/csharpstandard/pull/795/files/2976e80dce3b8349533c42521e1eb9c887105594#r1202295671) above.

```suggestion
- ***function-member***: Within a function a *variable_reference* to any of the following has a ref-safe-context of function-member:
  - Value parameters ([§9.2.5](variables.md#925-value-parameters)) on a function member declaration, including the implicit `this` of class member functions; and
  - The implicit reference (`ref`) parameter ([§9.2.6](variables.md#925-reference-parameters)) `this` of a struct member function, along with its fields.

   A *variable_reference* with ref-safe-context of function-member is a valid referent only if the reference variable is declared in the same function member.
- ***caller-context***:  Within a function a *variable_reference* to any of the following has a ref-safe-context of caller-context:
  - Reference (`ref`) parameters ([§9.2.6](variables.md#925-reference-parameters)) other than the implicit `this` of a struct member function;
  - Member fields and elements of such parameters;
  - Member fields of parameters of class type; and
  - Elements of parameters of array type.

  A *variable_reference* with ref-safe-context of caller-context can be the referent of a reference return. A *variable_reference* that can be the referent of a reference-return is ***safe-to-ref-return***.
```

I used nested bullets as the paragraphs were getting too dense. In case the suggestion box does not format well in your browser here it “unboxed”:

- ***function-member***: Within a function a *variable_reference* to any of the following has a ref-safe-context of function-member:
  - Value parameters ([§9.2.5](variables.md#925-value-parameters)) on a function member declaration, including the implicit `this` of class member functions; and
  - The implicit reference (`ref`) parameter ([§9.2.6](variables.md#925-reference-parameters)) `this` of a struct member function, along with its fields.

   A *variable_reference* with ref-safe-context of function-member is a valid referent only if the reference variable is declared in the same function member.
- ***caller-context***:  Within a function a *variable_reference* to any of the following has a ref-safe-context of caller-context:
  - Reference (`ref`) parameters ([§9.2.6](variables.md#925-reference-parameters)) other than the implicit `this` of a struct member function;
  - Member fields and elements of such parameters;
  - Member fields of parameters of class type; and
  - Elements of parameters of array type.

  A *variable_reference* with ref-safe-context of caller-context can be the referent of a reference return. A *variable_reference* that can be the referent of a reference-return is ***safe-to-ref-return***.

>
> ```csharp
> public class C
> {
>     // ref safe context of arr is "caller-context". 
>     // ref safe context of arr[i] is "caller-context".
>     private int[] arr = { 0, 1, 2, 3, 4, 5, 6, 7, 8, 9 }; 
> 
>     // ref safe context is "caller-context"
>     public ref int M1(ref int r1)
>     {
>         return ref r1; // r1 is safe to ref return
>     }
>
>     // ref safe context is "function-member"
>     public ref int M2(int v1)
>     {
>         return ref v1; // error: v1 isn't safe to ref return
>     }
>
>     public ref int M3()
>     {
>         int v2 = 5;
> 
>         return ref arr[v2]; // arr[v2] is safe to ref return
>     }
> 
>     public void M4(int p) 
>     {
>         int v3 = 6;
>
>         // context of r2 is block, ref safe context of p is method
>         ref int r2 = ref p;
>
>         // context of r3 is block, ref safe context of v3 is block
>         ref int r3 = ref v3;
>
>         // context of r4 is block, ref safe context of arr[v3] is caller method
>         ref int r4 = ref arr[v3]; 

> >     }

Reviewer 4: >} [](http://example.com/codeflow?start=6&length=1)

Should the comments within `M4()` use the exact terms defined above: "declaration-block", "caller-context", and "function-member", rather than block, caller method, and method?

> }
> ```
>
> *end example.*

#### §ref-span-safety-locals Local variable ref safe context

For a local variable `v`:

- If `v` is a reference variable, its ref-safe-context is the same as the ref-safe-context of its initializing expression.
- Otherwise its ref-safe-context is the context in which it was declared.

#### §ref-span-safety-parameters Parameter ref safe context

For a formal parameter `p`:

- If `p` is a `ref`, or `in` parameter, its ref-safe-context is the caller-context. It is safe-to-ref-return. If `p` is an `in` parameter, it can't be returned as a writable `ref` but can be returned as `ref readonly`.

> - If `p` is an `out` parameter, its ref-safe-context is the function-member. It isn't safe-to-ref-return.

Reviewer 4: >If `p` is an `out` parameter, its ref-safe-context is the function-member. [](http://example.com/codeflow?start=2&length=74)

In C#7, the ref-safe-context of an `out` parameter is caller-context since it can be returned by `ref`. (In C#11, `out` parameters cannot be returned by `ref`.)

Reviewer 2: A reminder to us all to always use an old version of the compiler for validation! I've added an [issue](https://github.com/dotnet/csharpstandard/issues/797) to remind (whoever is then) TG2 to change this “back” when C# 11 is reached – its a fairly obscure little change. Thanks @cston

## 9.6 Atomicity of variable references

Reads and writes of the following data types shall be atomic: `bool`, `char`, `byte`, `sbyte`, `short`, `ushort`, `uint`, `int`, `float`, and reference types. In addition, reads and writes of enum types with an underlying type in the previous list shall also be atomic. Reads and writes of other types, including `long`, `ulong`, `double`, and `decimal`, as well as user-defined types, need not be atomic. Aside from the library functions designed for that purpose, there is no guarantee of atomic read-modify-write, such as in the case of increment or decrement.

## §ref-span-safety Reference variables and returns

> ### §ref-span-safety-general General

Reviewer 5: I'm having a hard time reviewing this due to the change of terminology.  Then I realized all I had to do was... https://gist.github.com/gafter/d6068f35deb2a558b08b478f9dba1b1f (part 2...)

I'm sorry I ran out of time, but I think it would be a good idea to go through that gist and check off the rules one by one when we can find the corresponding rule in this PR.

Author: Yes, that's a good exercise. I'm confident all the rules are included. When I added this section, I started with this section, and then applied the following edits:

1. Re-order local and parameter rules.
2. Change title of "method invocation" to "function invocation" because the same rules would apply to local function declarations.
3. "Rvalue" changes to "Value" for consistency with the standard's existing terms.
4. "Span constructor" and "default" rules move to structs.md, in the section of "safe context" for `ref struct` types. 

I did a quick scan, and none of the rules were removed, or had other semantic edits.

Reviewer: Leaving this open to find an owner for going through the gist.

- Otherwise, if `p` is the `this` parameter of a struct type, its ref-safe-context is the function-member.
- Otherwise, the parameter is a value parameter, and its ref-safe-context is the function-member. It isn't safe-to-ref-return.

#### §ref-span-safety-field-reference Field ref safe context

For a variable designating a reference to a field, `e.F`:

- If `e` is of a reference type, its ref-safe-context is the caller-context.
- Otherwise, if `e` is of a value type, its ref-safe-context is the same as the ref-safe-context of `e`.

#### §ref-span-safety-operators Operators

The conditional operator (§12.18), `c ? ref e1 : ref e2`, and reference assignment operator, `= ref e` (§12.21.1) have reference variables as operands and yield a reference variable. For those operators, the ref-safe-context of the result is the narrowest context among the ref-safe-contexts of all `ref` operands.

#### §ref-span-safety-function-invocation Function invocation

For a variable `c` resulting from a ref-returning function invocation, `ref e1.M(e2, ...)`, its ref-safe-context is the narrowest of the following contexts:

- The caller-context.

> - The ref-safe-context of all `ref` and `out` argument expressions (excluding the receiver).

Reviewer 5: What about `in` arguments?

Author: `in` arguments are covered in the next line. The text was taken from https://gist.github.com/gafter/d6068f35deb2a558b08b478f9dba1b1f#method-invocation

Reviewer: No change needed.

> - For each `in` parameter of the method, if there is a corresponding expression that is a variable, its ref-safe-context, otherwise the nearest enclosing context

Reviewer: ```suggestion
- For each `in` parameter of the method, if there is a corresponding expression that is a variable, its ref-safe-context, otherwise the nearest enclosing context.
```
